pandas/core/indexes/base.py

```diff
@@ -251,6 +252,10 @@ def __new__(cls, data=None, dtype=None, copy=False, name=None,
         if name is None and hasattr(data, 'name'):
             name = data.name
 
+        if name is not None and not is_hashable(name):
```

Comment:
I don't actually think you need this here, as ``_set_name`` gets called from ``_simple_new`` (when ``.name`` is set). so rather more logical to put this in ``_set_name`` (like you do for MI)

Reply:
It seems like `_simple_new` calls Index's `__new__` through `result = object.__new__(cls)`. So if we delete this we won't have a hashability check on Index **construction**? 

For example `pd.Index([1, 2, 3], name=['foo'])` won't raise anymore. Is this ok?

Also note we are already doing checks in `_set_names` (for both Index & MI).

Reply:
Is there any way we can prevent the user from setting a non-hashable name **after** Index is created and if they do it **directly**  (i.e. not using `rename`, `set_names`,etc)? So that this will raise:
```
I1 = pd.Index([1, 2, 3], name='0')
I1.name = [[0]]

I1
>>> Int64Index([1, 2, 3], dtype='int64', name=[[0]])
```

Reply:
For MI, the name validation is delegated to `_set_names`:

https://github.com/pandas-dev/pandas/blob/2794474334cbd53315c248b605953579d010c693/pandas/core/indexes/multi.py#L223-L231

Can we do the same for Index here? Maybe something like this added in Index’s ` __new__()`:

```
_result = object.__new__(Index)

if name is not None:
    # handles name validation
    _result._set_names([name])
```

This way `pd.Index([1, 2, 3], name=['foo'])` would still raise but we do the check in a more logical place.